Python/Product/Analysis/LanguageServer/ParseQueue.cs

```diff
@@ -0,0 +1,198 @@
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.PythonTools.Analysis.Infrastructure;
+using Microsoft.PythonTools.Intellisense;
+using Microsoft.PythonTools.Parsing;
+using Microsoft.PythonTools.Parsing.Ast;
+
+namespace Microsoft.PythonTools.Analysis.LanguageServer {
+    class ParseQueue {
+        public const string PythonParserSource = "Python";
+        private const string TaskCommentSource = "Task comment";
+
+        private readonly ConcurrentDictionary<IDocument, Task<IAnalysisCookie>> _parsing;
+
+        private readonly VolatileCounter _parsingInProgress;
+
+        public ParseQueue() {
+            _parsingInProgress = new VolatileCounter();
+            _parsing = new ConcurrentDictionary<IDocument, Task<IAnalysisCookie>>();
+        }
+
+        public int Count => _parsingInProgress.Count;
+
+        public Task WaitForAllAsync() => _parsingInProgress.WaitForZeroAsync();
+
+        private static void AbortParsingTree(IPythonProjectEntry entry) {
+            if (entry == null) {
+                return;
+            }
+
+            entry.GetTreeAndCookie(out var tree, out var cookie);
+            entry.UpdateTree(tree, cookie);
+        }
+
+        public Task<IAnalysisCookie> Enqueue(IDocument doc, PythonLanguageVersion languageVersion) {
+            if (doc == null) {
+                throw new ArgumentNullException(nameof(doc));
+            }
+
+            (doc as IPythonProjectEntry)?.BeginParsingTree();
+            _parsingInProgress.Increment();
+
+            Task<IAnalysisCookie> result = null;
+            try {
+                result = _parsing.AddOrUpdate(doc,
+                    _ => Parse(doc, languageVersion, null),
+                    (_, prev) => Parse(doc, languageVersion, prev)
+                );
+                return result;
+            } finally {
+                if (result == null) {
+                    AbortParsingTree(doc as IPythonProjectEntry);
+                    _parsingInProgress.Decrement();
+                }
+            }
+        }
+
+        private Task<IAnalysisCookie> Parse(IDocument doc, PythonLanguageVersion languageVersion, Task<IAnalysisCookie> previousTask) {
+            if (previousTask != null && previousTask.Status == TaskStatus.Running) {
+                var callingThread = Thread.CurrentThread;
+                var res = previousTask.ContinueWith(t => {
+                    // We need to cancel immediately if we're running on the same thread
+                    // This can happen if the task completes before we call ContinueWith,
+                    // though the RunContinuationsAsynchronously should protect against this too
+                    if (Thread.CurrentThread == callingThread) {
```

Comment:
Delegate passed to `ContinueWith` may be executed in the caller thread only if `TaskContinuationOptions.ExecuteSynchronously` is explicitly passed to `ContinueWith`. Otherwise, task will be queued into scheduler. Otherwise, you can safely call `ContinueWith` for completed task.

Reply:
The debugger disagrees with you, though with the RunContinuationsAsynchronously it seems to work. Still, better to have the safety net in there than deal with bug reports later on (which I can guarantee will be vague performance related issues)

Reply:
> The debugger disagrees with you

That's something. Any specific repro?

Reply:
Ok, I was able to get an exception when `Thread.CurrentThread == callingThread`. There is no `ParseQueue.Parse` in the call stack, just continuation executed in the same `ThreadPool` thread the task with `ParseQueue.Parse` was executed before. This is normal - `ThreadPoolTaskScheduler` doesn't start new thread for each task, they are reused.

I don't have a repro, but here is what I think has happened:
1. `previousTask.Status == TaskStatus.Running` was true
2. `ParseQueue.Parse` has returned `res` (which wasn't canceled at that moment)
3. `res` was saved into `_parsing` and returned from `ParseQueue.Enqueue`
4. Since task is canceled, `ParseWorker` and `ProjectEntry.UpdateTree` won't be executed
5. `AnalysisQueue.Worker` will wait for `_workEvent` forever. This is the reasons (or at least one of the reasons) for the hanging @huguesv has found.

Reply:
So, if you have seen a continuation delegate and `ParseQueue.Parse` being in the same call stack without `TaskContinuationOptions.ExecuteSynchronously` explicitly specified, I believe it should be filed as a bug in Task implementation, cause such behavior is unexpected (synchronously executed continuation is an ad-hoc scenario).

If it is just the thread ids that match, then I think this method should be reduced to 
```cs
        private Task<IAnalysisCookie> Parse(IDocument doc, PythonLanguageVersion languageVersion, Task<IAnalysisCookie> previousTask) 
            => previousTask?.ContinueWith(t => ParseWorker(doc, languageVersion), CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default)
               ?? Task.Factory.StartNew(() => ParseWorker(doc, languageVersion), CancellationToken.None, TaskCreationOptions.None, TaskScheduler.Default);
```

`TaskScheduler.Default` ensures that task won't be inlined because of some unexpected logic in the `TaskScheduler.Current`. You can use `TaskCreationOptions.RunContinuationsAsynchronously` and `TaskContinuationOptions.RunContinuationsAsynchronously` to ensure that **all** continuations of these tasks are executed as separate thread pool work items.